ledger/ledgercore/accountdata.go

```diff
@@ -0,0 +1,180 @@
+	VoteFirstValid  basics.Round
+	VoteLastValid   basics.Round
+	VoteKeyDilution uint64
+
+	// MicroAlgosWithReward basics.MicroAlgos
+}
+
+// ToAccountData returns apply.AccountData from basics.AccountData
+func ToAccountData(acct basics.AccountData) AccountData {
+	return AccountData{
+		AccountBaseData: AccountBaseData{
+			Status:             acct.Status,
+			MicroAlgos:         acct.MicroAlgos,
+			RewardsBase:        acct.RewardsBase,
+			RewardedMicroAlgos: acct.RewardedMicroAlgos,
+
+			AuthAddr: acct.AuthAddr,
+
+			TotalAppSchema:      acct.TotalAppSchema,
+			TotalExtraAppPages:  acct.TotalExtraAppPages,
+			TotalAssetParams:    uint32(len(acct.AssetParams)),
+			TotalAssets:         uint32(len(acct.Assets)),
+			TotalAppParams:      uint32(len(acct.AppParams)),
+			TotalAppLocalStates: uint32(len(acct.AppLocalStates)),
+		},
+		VotingData: VotingData{
+			VoteID:          acct.VoteID,
+			SelectionID:     acct.SelectionID,
+			VoteFirstValid:  acct.VoteFirstValid,
+			VoteLastValid:   acct.VoteLastValid,
+			VoteKeyDilution: acct.VoteKeyDilution,
+		},
+	}
+}
+
+// AssignAccountData assigns the contents of AccountData to the fields in basics.AccountData,
+// but does not touch the AppParams, AppLocalState, AssetHolding, or AssetParams data.
+func AssignAccountData(a *basics.AccountData, acct AccountData) {
+	a.Status = acct.Status
+	a.MicroAlgos = acct.MicroAlgos
+	a.RewardsBase = acct.RewardsBase
+	a.RewardedMicroAlgos = acct.RewardedMicroAlgos
+
+	a.VoteID = acct.VoteID
+	a.SelectionID = acct.SelectionID
+	a.VoteFirstValid = acct.VoteFirstValid
+	a.VoteLastValid = acct.VoteLastValid
+	a.VoteKeyDilution = acct.VoteKeyDilution
+
+	a.AuthAddr = acct.AuthAddr
+	a.TotalAppSchema = acct.TotalAppSchema
+	a.TotalExtraAppPages = acct.TotalExtraAppPages
+}
+
+// WithUpdatedRewards calls basics account data WithUpdatedRewards
+func (u AccountData) WithUpdatedRewards(proto config.ConsensusParams, rewardsLevel uint64) AccountData {
+	bad := basics.AccountData{
+		Status:             u.Status,
+		MicroAlgos:         u.MicroAlgos,
+		RewardsBase:        u.RewardsBase,
+		RewardedMicroAlgos: u.RewardedMicroAlgos,
+	}
+	bad = bad.WithUpdatedRewards(proto, rewardsLevel)
+
+	u.MicroAlgos = bad.MicroAlgos
+	u.RewardsBase = bad.RewardsBase
+	u.RewardedMicroAlgos = bad.RewardedMicroAlgos
+	return u
+}
+
+// ClearOnlineState resets the account's fields to indicate that the account is an offline account
+func (u *AccountData) ClearOnlineState() {
+	u.Status = basics.Offline
+	u.VoteFirstValid = basics.Round(0)
+	u.VoteLastValid = basics.Round(0)
+	u.VoteKeyDilution = 0
+	u.VoteID = crypto.OneTimeSignatureVerifier{}
+	u.SelectionID = crypto.VRFVerifier{}
+}
+
+// MinBalance computes the minimum balance requirements for an account based on
+// some consensus parameters. MinBalance should correspond roughly to how much
+// storage the account is allowed to store on disk.
+func (u AccountData) MinBalance(proto *config.ConsensusParams) (res basics.MicroAlgos) {
```

Comment:
please extract the logic to be implemented in userBalance.go, and have it called from here.
i.e. pass all the relevant parameters on the stack. The compiler is going to inline these.

Reply:
done